File: PDOO/NapakalakiRuby.ruby/lib/numeric_bad_consequence.rb
#encoding: utf-8

require_relative 'bad_consequence'

module NapakalakiGame
  
  class NumericBadConsequence < BadConsequence
    
    def initialize(atext, somelevels, nVisible, nHidden)
      bc = super(atext, somelevels)
      @nVisibleTreasures=nVisible
      @nHiddenTreasures=nHidden
    end
  
    def getNVisibleTreasures
      @nVisibleTreasures
    end
  
    def getNHiddenTreasures
      @nHiddenTreasures
    end
    
    def isEmpty
    if @nVisibleTreasures == 0 && @nHiddenTreasures == 0 
        empty = true
      else
        empty = false
      end
      return empty
    end
    
    def substractVisibleTreasures(t)
      if @nVisibleTreasures > 0 then
        @nVisibleTreasures -= 1
      end
    end 
    
    def substractHiddenTreasures(t)
      if @nHiddenTreasures > 0 then
        @nHiddenTreasures -= 1
      end
    end
    
    public
    
    def adjustToFitTreasureLists(v, h)
=begin      
      if((@nVisibleTreasures)  < ( v.size + 1))
        nVT = @nVisibleTreasures
      else
        nVT = v.size
      end

      if (@nHiddenTreasures < h.size + 1)
        nHT = @nHiddenTreasures 
      else 
        nHT = h.size
      end
     
      if (isEmpty == false)
        NumericBadConsequence.new(getText, 0, nVT, nHT)
      end
=end
      n_visible = @nVisibleTreasures
      n_hidden = @nHiddenTreasures
      if v.size < n_visible then
        n_visible = v.size
      end
      if h.size < n_hidden then
      n_hidden = h.size
      end
      bc = NumericBadConsequence.new(@text, @levels, n_visible, n_hidden)
      bc
    end
    
    
    
    def to_s
      cadena = super
      cadena
    end 
  end
end

File: PDOO/Napakalaki.java/src/NapakalakiGame/NumericBadConsequence.java
package NapakalakiGame;

import java.util.ArrayList;

/**
 *
 * @author Andres Arco Lopez y Miguel Angel Cantarero Lopez
 */
public class NumericBadConsequence extends BadConsequence{
    static final int MAXTREASURES = 10;
    
    private int nVisibleTreasures;
    private int nHiddenTreasures;
    
    public NumericBadConsequence(String text, int level, int nVis, int nHid){
        super(text, level);
        nVisibleTreasures = nVis;
        nHiddenTreasures = nHid;
    }
    
    @Override
    public int getNVisibleTreasures(){
        return nVisibleTreasures;
    }
    
    @Override
    public int getNHiddenTreasures(){
        return nHiddenTreasures;
    }
    
    @Override 
    public ArrayList<TreasureKind> getSpecificVisibleTreasures(){
        return new ArrayList();
    }
    
    @Override
    public ArrayList<TreasureKind> getSpecificHiddenTreasures(){
        return new ArrayList();
    }
    
    @Override
    public boolean getDeath(){
        return false;
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t){
        if(getNVisibleTreasures()>0)
            nVisibleTreasures--;
    }
    
    @Override
    public void substractHiddenTreasure(Treasure t){
        if(getNHiddenTreasures()>0)
            nHiddenTreasures--;
    }
    
    @Override
    public boolean isEmpty(){
        boolean vacio = false;
        
        if(getNVisibleTreasures()==0 && getNHiddenTreasures()==0)
            vacio=true;
        return vacio;
    }
    
    @Override
    public BadConsequence adjustToFitTreasureLists(ArrayList <Treasure> v, ArrayList <Treasure> h){
        int nVis=nVisibleTreasures;
        int nHid=nHiddenTreasures;
        
        if(v.size()<nVis)
            nVis = v.size();
        if(h.size()<nHid)
            nHid = h.size();
        
        BadConsequence badcons = new NumericBadConsequence(text, levels, nVis, nHid);
        
        return badcons;
    }
    
    @Override
    public String toString(){
        String mensaje = super.toString();
        mensaje += "\nTesoros visibles perdidos: " + Integer.toString(nVisibleTreasures)
        +"\nTesoros ocultos perdidos: " + Integer.toString(nHiddenTreasures);
        
        return mensaje;
    }
}
